File: scripts/release/versioning.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const SCHEMA_VERSION = 1;

function assertWholeNumber(value, label) {
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`Invalid ${label}: ${value}`);
  }
}

function normalizeVersionState(input) {
  if (input == null || typeof input !== "object" || Array.isArray(input)) {
    throw new Error("Release version file must be a JSON object.");
  }

  if (input.schema !== SCHEMA_VERSION) {
    throw new Error(
      `Unsupported release version schema: ${String(input.schema)}. Expected ${SCHEMA_VERSION}.`,
    );
  }

  const current = input.current;
  if (current == null || typeof current !== "object" || Array.isArray(current)) {
    throw new Error("Release version file is missing a valid current version.");
  }

  const major = Number(current.major);
  const minor = Number(current.minor);
  const patch = Number(current.patch);

  assertWholeNumber(major, "major version");
  assertWholeNumber(minor, "minor version");
  assertWholeNumber(patch, "patch version");

  return {
    schema: SCHEMA_VERSION,
    current: {
      major,
      minor,
      patch,
    },
  };
}

export function formatVersion(version) {
  return `${version.major}.${version.minor}.${version.patch}`;
}

export async function readVersionState(filePath) {
  const raw = await fs.readFile(filePath, "utf8");
  return normalizeVersionState(JSON.parse(raw));
}

async function writeVersionState(filePath, state) {
  const normalized = normalizeVersionState(state);
  const payload = `${JSON.stringify(normalized, null, 2)}\n`;
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, payload, "utf8");
  return normalized;
}

export async function allocateNextPatchInFile(filePath) {
  const current = await readVersionState(filePath);
  const next = {
    schema: SCHEMA_VERSION,
    current: {
      major: current.current.major,
      minor: current.current.minor,
      patch: current.current.patch + 1,
    },
  };

  const saved = await writeVersionState(filePath, next);
  return {
    state: saved,
    version: formatVersion(saved.current),
  };
}

export async function setVersionBaseInFile(filePath, major, minor) {
  assertWholeNumber(major, "major version");
  assertWholeNumber(minor, "minor version");

  const next = {
    schema: SCHEMA_VERSION,
    current: {
      major,
      minor,
      patch: 0,
    },
  };

  const saved = await writeVersionState(filePath, next);
  return {
    state: saved,
    version: formatVersion(saved.current),
  };
}

function parseArgs(argv) {
  const [command, ...rest] = argv;
  const options = {};

  for (let index = 0; index < rest.length; index += 1) {
    const token = rest[index];
    if (!token.startsWith("--")) {
      throw new Error(`Unexpected argument: ${token}`);
    }
    const key = token.slice(2);
    const value = rest[index + 1];
    if (value == null || value.startsWith("--")) {
      throw new Error(`Missing value for --${key}`);
    }
    options[key] = value;
    index += 1;
  }

  return { command, options };
}

async function runCli() {
  const { command, options } = parseArgs(process.argv.slice(2));
  const filePath = options.file;

  if (!filePath) {
    throw new Error("Missing required --file argument.");
  }

  switch (command) {
    case "show": {
      const state = await readVersionState(filePath);
      process.stdout.write(`${JSON.stringify({
        state,
        version: formatVersion(state.current),
      })}\n`);
      return;
    }
    case "allocate": {
      const result = await allocateNextPatchInFile(filePath);
      process.stdout.write(`${JSON.stringify(result)}\n`);
      return;
    }
    case "set-base": {
      const major = Number(options.major);
      const minor = Number(options.minor);
      const result = await setVersionBaseInFile(filePath, major, minor);
      process.stdout.write(`${JSON.stringify(result)}\n`);
      return;
    }
    default:
      throw new Error(`Unsupported command: ${String(command)}`);
  }
}

const isDirectExecution = process.argv[1]
  ? path.resolve(process.argv[1]) === path.resolve(fileURLToPath(import.meta.url))
  : false;

if (isDirectExecution) {
  runCli().catch((error) => {
    process.stderr.write(`${error.message}\n`);
    process.exitCode = 1;
  });
}

File: scripts/release/release_lib/versioning.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


SCHEMA_VERSION = 1


def _assert_whole_number(value: Any, label: str) -> int:
    if not isinstance(value, int) or value < 0:
        raise ValueError(f"Invalid {label}: {value}")
    return value


def normalize_version_state(payload: dict[str, Any]) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise ValueError("Release version file must be a JSON object.")

    schema = payload.get("schema")
    if schema != SCHEMA_VERSION:
        raise ValueError(
            f"Unsupported release version schema: {schema}. Expected {SCHEMA_VERSION}.",
        )

    current = payload.get("current")
    if not isinstance(current, dict):
        raise ValueError("Release version file is missing a valid current version.")

    major = _assert_whole_number(current.get("major"), "major version")
    minor = _assert_whole_number(current.get("minor"), "minor version")
    patch = _assert_whole_number(current.get("patch"), "patch version")

    return {
        "schema": SCHEMA_VERSION,
        "current": {
            "major": major,
            "minor": minor,
            "patch": patch,
        },
    }


def format_version(version: dict[str, int]) -> str:
    return f"{version['major']}.{version['minor']}.{version['patch']}"


def read_version_state(file_path: str | Path) -> dict[str, Any]:
    path = Path(file_path)
    return normalize_version_state(json.loads(path.read_text(encoding="utf-8")))


def write_version_state(file_path: str | Path, state: dict[str, Any]) -> dict[str, Any]:
    normalized = normalize_version_state(state)
    path = Path(file_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(normalized, indent=2) + "\n", encoding="utf-8")
    return normalized


def allocate_next_patch_in_file(file_path: str | Path) -> dict[str, Any]:
    current = read_version_state(file_path)
    next_state = {
        "schema": SCHEMA_VERSION,
        "current": {
            "major": current["current"]["major"],
            "minor": current["current"]["minor"],
            "patch": current["current"]["patch"] + 1,
        },
    }
    saved = write_version_state(file_path, next_state)
    return {"state": saved, "version": format_version(saved["current"])}


def set_version_base_in_file(file_path: str | Path, major: int, minor: int) -> dict[str, Any]:
    next_state = {
        "schema": SCHEMA_VERSION,
        "current": {
            "major": _assert_whole_number(major, "major version"),
            "minor": _assert_whole_number(minor, "minor version"),
            "patch": 0,
        },
    }
    saved = write_version_state(file_path, next_state)
    return {"state": saved, "version": format_version(saved["current"])}
